File: Intatis-Windows/src/Intatis.Core/ChatLoop.cs
using Intatis.Core.Protocol;
using Intatis.Core.Providers;
using Intatis.Core.Session;

namespace Intatis.Core;

/// <summary>
/// Tool-free chat turn engine: durable user message, streamed assistant deltas into
/// the EventLog, terminal message_completed + turn_stats + turn_outcome. Chat has no
/// tools; every tool-capable surface goes through the AgentLoop instead.
/// </summary>
public sealed class ChatLoop
{
    private readonly EventLog _log;
    private readonly IChatProvider _provider;
    private readonly string _model;
    private readonly string? _systemPrompt;
    private readonly ReasoningEffort? _reasoningEffort;
    private readonly bool _includeUsage;

    public ChatLoop(
        EventLog log,
        IChatProvider provider,
        string model,
        string? systemPrompt = null,
        ReasoningEffort? reasoningEffort = null,
        bool includeUsage = false)
    {
        _log = log;
        _provider = provider;
        _model = model;
        _systemPrompt = systemPrompt;
        _reasoningEffort = reasoningEffort;
        _includeUsage = includeUsage;
    }

    /// <summary>Returns the terminal event sequence for this turn.</summary>
    public async Task<long> SendAsync(
        string userText,
        List<ImageAttachment>? images = null,
        CancellationToken ct = default)
    {
        var turnId = TurnId.New();
        var submissionId = SubmissionId.New();
        var messageId = MessageId.New();

        // Build history from the durable projection BEFORE appending this turn's
        // user message; the current user text is added explicitly below.
        var priorMessages = BuildHistory();
        var messages = new List<ChatMessage>(priorMessages)
        {
            new() { Role = "user", Content = userText, Images = images ?? [] },
        };

        _log.Append(EventType.UserMessage, new UserMessagePayload
        {
            Text = userText,
            SubmissionId = submissionId,
            TurnId = turnId,
            Attachments = images is { Count: > 0 }
                ? images.Select(_ => ArtifactId.New().Value).ToList()
                : null,
        }.ToJson());

        try
        {
            var usage = Usage.Empty;
            var text = "";
            await foreach (var chunk in _provider.StreamChatAsync(new ChatRequest
                           {
                               Model = _model,
                               Messages = messages,
                               ReasoningEffort = _reasoningEffort,
                               IncludeUsage = _includeUsage,
                           }, ct).ConfigureAwait(false))
            {
                ct.ThrowIfCancellationRequested();
                switch (chunk)
                {
                    case ChatChunk.Delta delta:
                        text += delta.Text;
                        _log.Append(EventType.MessageDelta, new MessageDeltaPayload
                        {
                            MessageId = messageId,
                            Role = MessageRoleWire.Assistant.ToWire(),
                            TextDelta = delta.Text,
                        }.ToJson(), flush: false);
                        break;

                    case ChatChunk.UsageReport report:
                        usage = usage.MergedWith(report.Usage);
                        break;
                }
            }

            _log.Append(EventType.MessageCompleted, new MessageCompletedPayload
            {
                MessageId = messageId,
                Role = MessageRoleWire.Assistant.ToWire(),
                Text = text,
            }.ToJson());
            _log.Append(EventType.TurnStats, new TurnStatsPayload
            {
                PromptTokens = usage.PromptTokens,
                CachedPromptTokens = usage.CachedPromptTokens,
                CompletionTokens = usage.CompletionTokens,
                TotalTokens = usage.TotalTokens,
                Model = _model,
            }.ToJson(), flush: false);
            var terminal = _log.Append(EventType.TurnOutcome, new TurnOutcomePayload
            {
                TurnId = turnId,
                Outcome = TurnOutcomeWire.Completed.ToWire(),
            }.ToJson());
            return terminal.Seq;
        }
        catch (OperationCanceledException)
        {
            // Never finalize a user-stopped partial into message_completed.
            _log.Append(EventType.TurnOutcome, new TurnOutcomePayload
            {
                TurnId = turnId,
                Outcome = TurnOutcomeWire.Interrupted.ToWire(),
                FailureSource = "turn_cancelled",
                Reason = "cancelled by user",
            }.ToJson());
            throw;
        }
        catch (Exception ex)
        {
            _log.Append(EventType.Error, new ErrorPayload
            {
                Code = "provider",
                Message = ex.Message,
            }.ToJson());
            var terminal = _log.Append(EventType.TurnOutcome, new TurnOutcomePayload
            {
                TurnId = turnId,
                Outcome = TurnOutcomeWire.Failed.ToWire(),
                FailureSource = "runtime_failed",
                Reason = ex.Message,
            }.ToJson());
            return terminal.Seq;
        }
    }

    private List<ChatMessage> BuildHistory()
    {
        // History from the canonical projection; only complete messages count.
        var projection = ConversationProjection.Build(_log.Replay());
        var messages = new List<ChatMessage>();
        if (_systemPrompt is { Length: > 0 })
            messages.Add(new ChatMessage { Role = "system", Content = _systemPrompt });
        foreach (var view in projection)
        {
            if (view.Role == MessageRoleWire.User)
                messages.Add(new ChatMessage { Role = "user", Content = view.Text });
            else if (view.Role is MessageRoleWire.Assistant or MessageRoleWire.Agent && view.IsComplete)
                messages.Add(new ChatMessage { Role = "assistant", Content = view.Text });
        }
        return messages;
    }
}

public static class TurnOutcomeWireExtensions
{
    public static string ToWire(this TurnOutcomeWire outcome) => outcome switch
    {
        TurnOutcomeWire.Completed => "completed",
        TurnOutcomeWire.Interrupted => "interrupted",
        _ => "failed",
    };
}

File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/ChatLoop.kt
package com.intatis.shared

import com.intatis.shared.protocol.ErrorPayload
import com.intatis.shared.protocol.EventType
import com.intatis.shared.protocol.MessageCompletedPayload
import com.intatis.shared.protocol.MessageDeltaPayload
import com.intatis.shared.protocol.MessageRole
import com.intatis.shared.protocol.TurnOutcomePayload
import com.intatis.shared.protocol.TurnOutcomeWire
import com.intatis.shared.protocol.TurnStatsPayload
import com.intatis.shared.protocol.UserMessagePayload
import com.intatis.shared.providers.ChatChunk
import com.intatis.shared.providers.ChatMessage
import com.intatis.shared.providers.ChatRequest
import com.intatis.shared.providers.ChatProvider
import com.intatis.shared.providers.ImageAttachment
import com.intatis.shared.providers.ReasoningEffort
import com.intatis.shared.providers.Usage
import com.intatis.shared.session.ConversationProjection
import com.intatis.shared.session.EventLog
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

/**
 * Tool-free chat turn engine (the mobile surface is a strict Chat subset per the
 * Apple contract): durable user message, streamed assistant deltas into the
 * EventLog, terminal message_completed + turn_stats + turn_outcome.
 */
class ChatLoop(
    private val log: EventLog,
    private val provider: ChatProvider,
    private val model: String,
    private val systemPrompt: String? = null,
    private val reasoningEffort: ReasoningEffort? = null,
    private val includeUsage: Boolean = false,
) {

    /** Returns the terminal event sequence for this turn. */
    suspend fun send(
        userText: String,
        images: List<ImageAttachment>? = null,
    ): Long {
        val turnId = TurnId.new()
        val submissionId = SubmissionId.new()
        val messageId = MessageId.new()

        // History from the canonical projection BEFORE appending this turn's user
        // message; the current user text is added explicitly below.
        val priorMessages = buildHistory()
        val messages = priorMessages.toMutableList()
        messages.add(ChatMessage("user", userText, images ?: emptyList()))

        log.append(EventType.USER_MESSAGE, UserMessagePayload(
            text = userText,
            submissionId = submissionId.value,
            turnId = turnId.value,
            attachments = if (!images.isNullOrEmpty()) {
                images.map { ArtifactId.new().value }
            } else null,
        ).toJson())

        return try {
            var usage = Usage.EMPTY
            val text = StringBuilder()

            provider.streamChat(ChatRequest(
                model = model,
                messages = messages,
                reasoningEffort = reasoningEffort,
                includeUsage = includeUsage,
            )).collect { chunk ->
                currentCoroutineContext().ensureActive()
                when (chunk) {
                    is ChatChunk.Delta -> {
                        text.append(chunk.text)
                        log.append(EventType.MESSAGE_DELTA, MessageDeltaPayload(
                            messageId = messageId.value,
                            role = MessageRole.ASSISTANT.wire,
                            textDelta = chunk.text,
                        ).toJson(), flush = false)
                    }

                    is ChatChunk.UsageReport -> usage = usage.mergedWith(chunk.usage)

                    ChatChunk.Done -> { /* terminal markers appended below */ }
                }
            }

            log.append(EventType.MESSAGE_COMPLETED, MessageCompletedPayload(
                messageId = messageId.value,
                role = MessageRole.ASSISTANT.wire,
                text = text.toString(),
            ).toJson())
            log.append(EventType.TURN_STATS, TurnStatsPayload(
                promptTokens = usage.promptTokens,
                cachedPromptTokens = usage.cachedPromptTokens,
                completionTokens = usage.completionTokens,
                totalTokens = usage.totalTokens,
                model = model,
            ).toJson(), flush = false)
            log.append(EventType.TURN_OUTCOME, TurnOutcomePayload(
                turnId = turnId.value,
                outcome = TurnOutcomeWire.COMPLETED.wire,
            ).toJson()).seq
        } catch (e: kotlinx.coroutines.CancellationException) {
            // Never finalize a user-stopped partial into message_completed.
            log.append(EventType.TURN_OUTCOME, TurnOutcomePayload(
                turnId = turnId.value,
                outcome = TurnOutcomeWire.INTERRUPTED.wire,
                failureSource = "turn_cancelled",
                reason = "cancelled by user",
            ).toJson())
            throw e
        } catch (e: Exception) {
            log.append(EventType.ERROR, ErrorPayload(
                code = "provider",
                message = e.message ?: "provider failure",
            ).toJson())
            log.append(EventType.TURN_OUTCOME, TurnOutcomePayload(
                turnId = turnId.value,
                outcome = TurnOutcomeWire.FAILED.wire,
                failureSource = "runtime_failed",
                reason = e.message ?: "provider failure",
            ).toJson()).seq
        }
    }

    private fun buildHistory(): List<ChatMessage> {
        val projection = ConversationProjection.build(log.replay())
        val messages = mutableListOf<ChatMessage>()
        systemPrompt?.takeIf { it.isNotEmpty() }?.let {
            messages.add(ChatMessage("system", it))
        }
        for (view in projection) {
            when {
                view.role == MessageRole.USER ->
                    messages.add(ChatMessage("user", view.text))

                (view.role == MessageRole.ASSISTANT || view.role == MessageRole.AGENT) && view.isComplete ->
                    messages.add(ChatMessage("assistant", view.text))
            }
        }
        return messages
    }
}
